src/continuous_agg.c

```diff
@@ -400,21 +294,60 @@ ts_continuous_agg_find_by_mat_hypertable_id(int32 mat_hypertable_id)
 }
 
 ContinuousAgg *
-ts_continuous_agg_find_by_view_name(const char *schema, const char *name)
+ts_continuous_agg_find_by_view_name(const char *schema, const char *name,
+									ContinuousAggViewType type)
 {
-	ScanIterator iterator =
-		ts_scan_iterator_create(CONTINUOUS_AGG, AccessShareLock, CurrentMemoryContext);
+	ScanIterator iterator;
 	ContinuousAgg *ca = NULL;
+	AttrNumber view_name_attrnum = 0;
+	AttrNumber schema_name_attrnum = 0;
 	int count = 0;
 
+	Assert(schema);
+	Assert(name);
+
+	switch (type)
+	{
+		case ContinuousAggUserView:
+			schema_name_attrnum = Anum_continuous_agg_user_view_schema;
+			view_name_attrnum = Anum_continuous_agg_user_view_name;
+			break;
+		case ContinuousAggPartialView:
+			schema_name_attrnum = Anum_continuous_agg_partial_view_schema;
+			view_name_attrnum = Anum_continuous_agg_partial_view_name;
+			break;
+		case ContinuousAggDirectView:
+			schema_name_attrnum = Anum_continuous_agg_direct_view_schema;
+			view_name_attrnum = Anum_continuous_agg_direct_view_name;
+			break;
+		case ContinuousAggAnyView:
+			break;
+	}
+
+	iterator = ts_scan_iterator_create(CONTINUOUS_AGG, AccessShareLock, CurrentMemoryContext);
+
+	if (type != ContinuousAggAnyView)
+	{
+		ts_scan_iterator_scan_key_init(&iterator,
+									   schema_name_attrnum,
+									   BTEqualStrategyNumber,
+									   F_NAMEEQ,
+									   CStringGetDatum(schema));
+		ts_scan_iterator_scan_key_init(&iterator,
+									   view_name_attrnum,
+									   BTEqualStrategyNumber,
+									   F_NAMEEQ,
+									   CStringGetDatum(name));
+	}
+
 	ts_scanner_foreach(&iterator)
 	{
 		bool should_free;
 		HeapTuple tuple = ts_scan_iterator_fetch_heap_tuple(&iterator, false, &should_free);
 		FormData_continuous_agg *data = (FormData_continuous_agg *) GETSTRUCT(tuple);
 		ContinuousAggViewType vtyp = ts_continuous_agg_view_type(data, schema, name);
```

Comment:
This line seems unnecessary to be called if `type != ContinuousAggAnyView`. I understand that it helps to reuse code, but it makes it difficult to understand the overall function logic. I expect that if `type != ContinuousAggAnyView` then `vtyp` will be always equivalent to `type`, right?

Reply:
You can call the function with `ContinuousAggAnyView`, which means match any view (out of the three). So, the returned vtyp is in most cases something else than `ContinuousAggAnyView`

Reply:
@ I will repeat my statement, since it looks like you are missing it:

If `type != ContinuousAggAnyView`, then calling `ContinuousAggViewType vtyp = ts_continuous_agg_view_type(data, schema, name)` will return the same type as in `type`.

Your answer is about when `type == ContinuousAggAnyView` and I agree that `vtyp` will be different, otherwise there will be never result.

Reply:
Does it work for you now?